services/supervisor/tests/supervisor/migrations/mod.rs

```diff
@@ -0,0 +1,879 @@
+        let builder = builder
+            .with_rust_service(MigrationServiceV05)
+            .with_artifact(MigrationServiceV05.artifact_id())
+            .with_instance(MigrationServiceV05.artifact_id().into_default_instance(
+                MigrationService::INSTANCE_ID,
+                MigrationService::INSTANCE_NAME,
+            ));
+
+        // Add migrating artifact for version 0.7.
+        let builder = builder
+            .with_migrating_rust_service(FailingMigrationServiceV07)
+            .with_artifact(FailingMigrationServiceV07.artifact_id());
+
+        builder.create()
+    };
+
+    // Stop service instance before running the migration.
+    stop_service(&mut testkit, MigrationService::INSTANCE_ID);
+
+    // Request migration.
+    let deadline_height = DEADLINE_HEIGHT;
+    let request = MigrationRequest {
+        new_artifact: FailingMigrationServiceV07.artifact_id(),
+        service: MigrationService::INSTANCE_NAME.into(),
+        deadline_height,
+    };
+
+    let api = testkit.api();
+    let tx_hash = request_migration(&api, request.clone());
+    let block = testkit.create_block();
+
+    block[tx_hash]
+        .status()
+        .expect("Transaction should be executed successfully");
+
+    let error = wait_for_migration_fail(&mut testkit, deadline_height, request);
+
+    assert_eq!(
+        error,
+        ErrorMatch::from_fail(&MigrationError::MigrationFailed)
+            .with_description_containing("This migration always fails")
+    );
+}
+
+/// This test checks that migration that contains two migration scripts completes
+/// successfully in two steps.
+#[test]
+fn complex_migration() {
+    let mut testkit = testkit_with_supervisor_and_service(1);
+
+    // Stop service instance before running the migration.
+    stop_service(&mut testkit, MigrationService::INSTANCE_ID);
+
+    // Request migration to 0.5.
+    // This migration will require two migration requests.
+    let deadline_height = DEADLINE_HEIGHT;
+    let request = MigrationRequest {
+        new_artifact: MigrationServiceV05.artifact_id(),
+        service: MigrationService::INSTANCE_NAME.into(),
+        deadline_height,
+    };
+
+    let api = testkit.api();
+    let tx_hash = request_migration(&api, request.clone());
+    let block = testkit.create_block();
+
+    block[tx_hash]
+        .status()
+        .expect("Transaction should be executed successfully");
+
+    // After the first migration step, version should be "0.2".
+    wait_for_migration_success(
+        &mut testkit,
+        deadline_height,
+        request,
+        Version::new(0, 2, 0),
+    );
+
+    let snapshot = testkit.snapshot();
+    let prefixed = Prefixed::new(MigrationService::INSTANCE_NAME, snapshot.as_ref());
+
+    migration_service::v02::verify_schema(prefixed);
+
+    // Request the same migration.
+    let deadline_height = Height(DEADLINE_HEIGHT.0 * 2);
+    let request = MigrationRequest {
```

Comment:
Nit: To better illustrate the point, you could modify the deadline height in the previous `request`.